alembic/script/revision.py

```diff
@@ -40,6 +40,34 @@ def __init__(self, message, argument):
         self.argument = argument
 
 
+class LoopDetected(RevisionError):
+    def __init__(self, revision):
+        self.revision = revision
+        super(LoopDetected, self).__init__(
+            "Self-loop is detected on revision %s" % (revision,)
+        )
+
+
+class DependencyLoopDetected(RevisionError):
+    def __init__(self, revision):
+        self.revision = revision
+        super(DependencyLoopDetected, self).__init__(
+            "Dependency self-loop is detected on revision %s" % (revision,)
+        )
+
+
```

Comment:
**mike bayer** (zzzeek) wrote:

is there a way we can make these exception classes in a deeper hierarchy?   that is:

1. is a "Loop" a specific case of "Cycle"?   in that case LoopDetected could extend CycleDetected

2. "Dependency" vs. non, is a DependencyLoop a subset of Loop?   DependencyCycle -> Cycle. 

3. or even DependencyLoop -> (Loop, DependencyCycle)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/alembic/+/2379